src/v/raft/coordinated_recovery_throttle.cc

```diff
@@ -21,46 +21,75 @@
 #include <boost/range/irange.hpp>
 
 #include <chrono>
+#include <cstdio>
+#include <iterator>
+#include <limits>
 
 namespace raft {
 
 coordinated_recovery_throttle::token_bucket::token_bucket(
   size_t initial_capacity)
-  : _sem(initial_capacity, "recovery_throttle")
-  , _last_reset_capacity(initial_capacity) {}
+  : _available_units(initial_capacity)
+  , _last_reset_capacity(initial_capacity) {
+    _cv.signal();
+}
 
 ss::future<> coordinated_recovery_throttle::token_bucket::throttle(
   size_t size, ss::abort_source& as) {
     _waiting_bytes += size;
     auto decrement = ss::defer([this, size] { _waiting_bytes -= size; });
-    co_await _sem.wait(as, size);
-    _admitted_bytes_since_last_reset += size;
+    auto lock = co_await _mutex.get_units(as);
```

Comment:
i am not sure i understand why this lock is required here, isn't it enough to wait on the condition variable ? Now all the throttle requests will be enqueued in the lock waiters queue, and only one of them is waiting for the `_cv` and check the predicates

Reply:
I think my comment was misunderstood.. I was saying we don't need the lock because CV already enforces the queue order, since it is only signaled in all other places (so only front of the queue runs the critical section). mutex seems like one additional layer of unneeded queueing unless I"m missing something.

Reply:
This is probably worth documenting in the header.